### PR TITLE
Define Content-Length parser

### PR DESCRIPTION
**Pull request by annevk**
As the body concept is refactored for #604, XMLHttpRequest will need to use this algorithm for its events. We also want to require browsers to use this algorithm instead of the one defined by HTTP as part of #1156.

Tests: https://github.com/web-platform-tests/wpt/pull/10548.

- [x] At least two implementers are interested (and none opposed):
   * Chrome
   * Firefox
- [x] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * See above.
- [x] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: N/A
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1695931 & https://bugzilla.mozilla.org/show_bug.cgi?id=1696203
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=222596

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

@MattMenke2 I haven't attempted to actually require browsers use this yet. I suspect we want some kind of "HTTP monkey patching" section that outlines the various things we want browsers (and clients that emulate them) to implement for that. Ideas welcome, but it's not my main priority here (I'm trying to scope my yak shaving).

***
<a href="https://whatpr.org/fetch/1183.html" title="Last updated on Mar 3, 2021, 4:18 PM UTC (78d9116)">Preview</a> | <a href="https://whatpr.org/fetch/1183/24a8670...78d9116.html" title="Last updated on Mar 3, 2021, 4:18 PM UTC (78d9116)">Diff</a>

### CODE REVIEW

**Comment by annevk** [comment]
One could argue that we should only define this for a response. I don't think I have found a place where browsers use this for requests.

**Review by MattMenke2 — APPROVED** [review]
This change looks good to me.

I'm not sure I'll have the time to fully dig through #1184, make sure it happily interacts with the chunked encoding bits, etc.

**Comment by annevk** [comment]
I've decided to land this part and leave landing the additional tests and browser bugs to #1184 as this is fairly standalone.

**Comment by mnot** [comment]
When I run http://wpt.live/fetch/content-length/content-length.html in both FF and Safari, my browser doesn't fetch `content-lengths.json`, and only runs one test; wpt.fyi shows only one test result at <https://wpt.fyi/results/fetch/content-length/content-length.html?label=master&label=experimental&aligned>. What am I missing?

**Comment by MattMenke2** [comment]
Hrm...You're right.  Looks like there's nothing that loads parsing.window.js or too-long.window.js.

Anne:  Did you forget to upload a file?

**Comment by annevk** [comment]
No, you want to run http://wpt.live/fetch/content-length/parsing.window.html and http://wpt.live/fetch/content-length/too-long.window.html for the added tests. `.window.js` (as well as `.any.js`) are shorthands to avoid having to write a bunch of boilerplate: https://web-platform-tests.org/writing-tests/testharness.html#without-html-boilerplate-window-js.

**Comment by mnot** [comment]
Aha - they are suddenly showing up now. Weird, they weren't before, even with the same commit; caching, maybe?

**Comment by annevk** [comment]
Ah, I think the results page is often behind a bit as it takes a while to run all the tests. So it only updates every so often.

**Comment by mnot** [comment]
Considering that you're really only showing interoperability between Blink-derived browsers, it's a bit premature to be saying that 'the model defined in HTTP is not compatible with web content', don't you think?

**Comment by annevk** [comment]
I suppose that Safari's results are closest to indicating that things could perhaps be more strict (although Safari is less strict than Chrome in many cases as well).

Perhaps what you're saying though is that https://httpwg.org/http-core/draft-ietf-httpbis-semantics-latest.html#rfc.section.8.6 doesn't contain many requirements on recipients so anything is fine? But even then, the model where recipients get to choose how to behave does not seem compatible as duplicate `Content-Length` headers are a thing.

**Comment by martinthomson** [comment]
The reporting here is just odd.  All the percent encoding, and no expectations.  Do these have to be like this?

**Comment by annevk** [comment]
There's a variety of ways to do it, but this seemed nice in that you can see the exact input (albeit encoded).

**Comment by martinthomson** [comment]
It doesn't say what the expectation is.  And all the percent encoding makes it very hard to parse.  Given that this test doesn't depend on the spacing, removing the encoding would be a start at making the output readable.

**Comment by annevk** [comment]
Sure: https://github.com/web-platform-tests/wpt/pull/27894. (The expectations are also part of the Message column, but it can be a bit cryptic if you don't know what you're looking for.)
